RavenDB-23815 Update install_build_prequisite.sh for Ubuntu 24.04 compatibility

Jay-Lokhande opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23815

### Additional description

This PR enhances the RavenDB build environment setup script with improved structure, better compatibility, and streamlined installations. The changes make the script more reliable, easier to maintain, and compatible with newer Ubuntu versions.

Key Changes:
✅ Added an introductory message to display Ubuntu version and codename for better clarity.
✅ Enforced root privileges check using $EUID to prevent issues with script execution.
✅ Expanded supported Ubuntu versions to include 22.04 and 24.04, ensuring wider compatibility.
✅ Replaced which with command -v for better POSIX compliance in command detection.
✅ Simplified Microsoft repository setup by using a dynamic URL, reducing redundant conditionals.
✅ Upgraded .NET SDK installation dynamically from global.json and ensured it initializes correctly.
✅ Added PowerShell installation check to avoid redundant installations.
✅ Improved Node.js installation by using the LTS version from Nodesource, ensuring compatibility.
✅ Ensured Git is installed before proceeding with the build.
✅ Refactored script output messages to provide clear guidance to users during setup.

Testing Done:
✔️ Tested on Ubuntu 20.04, 22.04, and 24.04.
✔️ Verified that .NET, PowerShell, Node.js, and Git install correctly.
✔️ Confirmed script runs smoothly without unnecessary re-installations when dependencies are already installed.

### Type of change

- [x] Bug fix
- [x] Regression bug fix
- [x] Optimization

### How risky is the change?

- [x] Low 

### Backward compatibility

- [x] Non breaking change

### Is it platform specific issue?

- [x] Yes. Please list the affected platforms.

### Documentation update

- [x] No documentation update is needed 

### Testing by Contributor

- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [x] No

### UI work

- [x] No UI work is needed

Jay-Lokhande commented:
I am working on the remaining issues that has encountered.

Jay-Lokhande commented:
The script works on both Ubuntu 24.04 and 22.04, but I’ve only added the configuration for 24.04. Do you think we should add it for 22.04 as well?

ayende commented:
It would be best, yes.

ayende commented on install_build_prerequisites.sh:
```diff
@@ -9,8 +9,9 @@ DOTNET_VERSION_CMD=`dotnet --version 2> /dev/null`
-    echo "Unsupported Ubuntu version: $UBUNTU_VERSION $UBUNTU_CODENAME. Must be 20.04, 18.04, 16.04 or 14.04."
+
+if [[ ! "$UBUNTU_VERSION" =~ ^(16|18|20|22|24)\.04$ ]] ; then
+    echo "Unsupported Ubuntu version: $UBUNTU_VERSION $UBUNTU_CODENAME. Must be 24.04, 20.04, 18.04, or 16.04."
```
```suggestion
    echo "Unsupported Ubuntu version: $UBUNTU_VERSION $UBUNTU_CODENAME. Must be 24.04, 22.04, 20.04, 18.04, or 16.04."
```

Jay-Lokhande committed:
RavenDb-23815 add introductory message displaying Ubuntu version and codename

ayende commented on install_build_prerequisites.sh:
```diff
@@ -1,45 +1,43 @@
 
-if [[ $UID != 0 ]]; then
-    echo "Please run this script with sudo:"
-    echo "sudo $0 $*"
```
This is important, it means that you can copy/paste the fixed command

Jay-Lokhande replied:
Hi, I've updated the script to use the EUID check instead of UID. EUID is more reliable in modern environments, especially when running under sudo.  or should we go back to UID.

ayende replied:
I meant, the `echo sudo...` line

ayende commented on install_build_prerequisites.sh:
```diff
@@ -1,55 +1,52 @@
+
+# Install .NET SDK
+echo "Installing .NET Core SDK 8.0..."
+apt-get install -y dotnet-sdk-8.0
```
Here we should probably consider the `https://github.com/ravendb/ravendb/blob/v7.0/global.json` file - since that changes and we don't want to have to update this script each time

Jay-Lokhande replied:
Got it! I’ve updated the script to read the .NET SDK version directly from global.json, so we don’t have to update it manually anymore. It now extracts the version, picks the right package, and installs it. Also added some error handling in case global.json is missing or incorrect.

Jay-Lokhande commented:
I need to add 'RavenDB-XXXXX:' before each commit message, but I'm not sure what 'XXXXX' should be. Could you please confirm if it refers to an issue number or provide the correct value?

ayende commented:
You need to create a new issue here: 
https://issues.hibernatingrhinos.com/issues/RavenDB?

That will give you the actual issue number for the commits

Jay-Lokhande committed:
RavenDB-23815 add introductory message displaying Ubuntu version and codename

Jay-Lokhande committed:
RavenDB-23815 enforce root privileges using EUID check

Jay-Lokhande committed:
RavenDB-23815 expand supported Ubuntu versions to include 16.04, 18.04, 20.04, 22.04, and 24.04

Jay-Lokhande committed:
RavenDB-23815 replace 'which' with 'command -v' for dependency checks

Improves portability by using 'command -v' instead of 'which' for checking the existence of required commands.

Jay-Lokhande committed:
RavenDB-23815 simplify Microsoft repository setup using dynamic URL based on Ubuntu version

Jay-Lokhande committed:
RavenDB-23815 dynamically extract and install .NET SDK version from global.json

- Added a check for jq and installed it if missing to enable JSON parsing.
- Verified that global.json exists in the repository root.
- Extracted the .NET SDK version from global.json using jq.
- Derived the major.minor version (e.g., 8.0 from 8.0.100) for apt-get compatibility.
- Installed the corresponding dotnet-sdk package based on the extracted version.
- Added error handling and verification steps for robust installation.

Jay-Lokhande committed:
RavenDB-23815 add check and installation for PowerShell

Jay-Lokhande committed:
RavenDB-23815 install Node.js using Nodesource LTS script and enforce minimum version

Jay-Lokhande committed:
RavenDB-23815 add check for Git and install if missing

Jay-Lokhande committed:
RavenDB-23815 add informative echo messages to guide users through the setup process

Jay-Lokhande commented:
I have made all the necessary changes. @gregolsky  @ayende , please review them. These changes are running successfully on my system.

Jay-Lokhande committed:
RavenDB-23815 Remove unwanted text from script

Jay-Lokhande commented:
Updated the script to automatically download the latest Node.js version if not found. If Node.js is already installed, it now checks whether the version is greater than 8 to determine compatibility. Considering that Node.js 8 is deprecated, should we update the check to require version 18 or 19 instead?

Jay-Lokhande commented:
Hey @ayende  @gregolsky ,

I’ve made some updates to the script as suggested by @ayende  and would love to get your feedback on the changes.

Jay-Lokhande commented:
> You need to create a new issue here: https://issues.hibernatingrhinos.com/issues/RavenDB?
> 
> That will give you the actual issue number for the commits

Hello Sir,

Apologies for not acknowledging earlier. I have created the issue as required: [23815] ([Issue Link](https://github.com/ravendb/ravendb/pull/20180#)). I have also made the necessary changes accordingly. Please let me know if any further modifications are needed.

Thank you!

Jay-Lokhande commented:
@gregolsky can you please review my pr actually earlier commits messages were not in standard format thats why I changed to new commits messages in required format with issue numbers.

Jay-Lokhande committed:
Update CONTRIBUTING.md to reflect .NET version from global.json

Jay-Lokhande committed:
Merge branch 'ravendb:v7.0' into ubuntu-24.04-support

Jay-Lokhande commented:
I updated CONTRIBUTING.md to specify the `.NET SDK` version from `global.json` instead of `.NET 6`. Additionally, I think we should mention running `install_build_prerequisite.sh` before` ./build.sh` to ensure all dependencies are installed. Let me know your thoughts!

ppekrol commented:
hi @Jay-Lokhande 
we will be looking into your changes this week, were a bit busy lately, thanks!

Jay-Lokhande commented:
> hi @Jay-Lokhande we will be looking into your changes this week, were a bit busy lately, thanks!

Thanks for the update! I appreciate your time and look forward to your feedback.

Jay-Lokhande commented:
Hey @ayende @gregolsky @ppekrol @arekpalinski ,
I just wanted to take a moment to sincerely thank you all for your guidance and support throughout my contributions. I truly appreciate the time and effort you’ve put into reviewing my work. Excited to keep collaborating!
